### PR TITLE
DOCSP-42957: DateTimeInterface in queries

### PR DESCRIPTION
**Pull request by norareidy**
JIRA - https://jira.mongodb.org/browse/DOCSP-42957
Staging - https://preview-mongodbnorareidy.gatsbyjs.io/laravel/DOCSP-42957-stage/upgrade/#version-5.x-breaking-changes

### Checklist

- [ ] Add tests and ensure they pass
- [ ] Add an entry to the CHANGELOG.md file

### CODE REVIEW

File: docs/upgrade.txt [hunk 1]
```diff
@@ -89,6 +89,9 @@ This library version introduces the following breaking changes:
   of this behavior, you cannot have two separate ``id`` and ``_id`` fields in your
   documents.
 
+- When using a ``DateTimeInterface`` object in a query, the library converts
+  the ``DateTimeInterface`` to a ``MongoDB\BSON\UTCDateTime`` object.
```
Comment (GromNaN) [inline docs/upgrade.txt]:
```suggestion
- When using a ``DateTimeInterface`` object (including ``Carbon``) in a query, the library converts
  the ``DateTimeInterface`` to a ``MongoDB\BSON\UTCDateTime`` object.
```

**Review by GromNaN — COMMENTED** [review]
The same applies in the other direction. UTCDateTime objects received as results are converted to `Carbon` date with the default timezone. https://github.com/mongodb/laravel-mongodb/pull/3119

I think this should be mentioned in the query builder docs, with an example.

File: docs/upgrade.txt (hunk 1 above)
Comment (rustagir) [inline docs/upgrade.txt]: S: It's a bit unclear what it means to use an object "in a query". Perhaps you could provide a short example (code or descriptive)

**Review by rustagir — CHANGES_REQUESTED** [review]
I left a suggestion that might change the scope of the PR, so requesting another look!

File: docs/upgrade.txt (hunk 1 above)
Comment (GromNaN) [reply docs/upgrade.txt]: Query refers to the filter passed to `Query\Builder::where()`, or the data to `insert()` or `update()`.

File: docs/includes/query-builder/QueryBuilderTest.php [hunk 2]
```diff
@@ -150,12 +150,13 @@ public function testWhereIn(): void
...
-        // begin query whereDate
+        // begin query where date
         $result = DB::connection('mongodb')
             ->table('movies')
-            ->whereDate('released', '2010-1-15')
-            ->get();
-        // end query whereDate
+            ->where('released', new UTCDateTime(
+                Carbon::create(2010, 1, 15, 0, 0, 0, 'UTC')
+            ))->get();
```
Comment (GromNaN) [inline docs/includes/query-builder/QueryBuilderTest.php]: The previous example was a lot better as it used `whereDate` method that convert the string to a date object. I don't think it makes sense to change it.

The purpose is to show that users can work with `Carbon` object without worring with `UTCDateTime`. This example shows the opposite of what should be done.

```php
->where('released', Carbon::create(2010, 1, 15));
```

File: docs/query-builder.txt [hunk 3]
```diff
@@ -348,23 +348,26 @@ query builder method to retrieve documents from the
 Match Dates Example
 ^^^^^^^^^^^^^^^^^^^
 
-The following example shows how to use the ``whereDate()``
+The following example shows how to use the ``where()``
 query builder method to retrieve documents from the
-``movies`` collection that match the specified date of
-``2010-1-15`` in the ``released`` field:
+``movies`` collection in which the ``released`` value
+is ``2010, 1, 15, 0, 0, 0``, specified in a ``Carbon`` object:
```
Comment (GromNaN) [inline docs/query-builder.txt]: The value is quoted without explaining its meanings. It's pointless. I wouldn't even put it there. The previous example was more explicit.

File: docs/upgrade.txt (hunk 1 above)
Comment (norareidy) [reply docs/upgrade.txt]: I added a brief description and linked to an example on the query builder page

**Review by GromNaN — CHANGES_REQUESTED** [review]
Adding a 2nd example would make more sense that changing the existing one.

File: docs/upgrade.txt [hunk 4]
```diff
@@ -81,6 +81,16 @@ This library version introduces the following breaking changes:
...
     
+- When using a ``DateTimeInterface`` object, including ``Carbon``, in a query,
+  the library converts the ``DateTimeInterface`` to a ``MongoDB\BSON\UTCDateTime``
+  object. This conversion applies to ``DateTimeInterface`` objects passed as query
+  filters to the ``where()`` method or as data passed to the ``insert()`` and
+  ``update()`` methods. 
+  
+  For an example that passes a ``Carbon`` object to the
+  ``DB::where()`` method, see the :ref:`laravel-query-builder-wheredate`
+  section of the Query Builder guide.
```
Comment (rustagir) [inline docs/upgrade.txt]:
```suggestion
  To view an example that passes a ``Carbon`` object to the
  ``DB::where()`` method, see the :ref:`laravel-query-builder-wheredate`
  section of the Query Builder guide.
```

File: docs/query-builder.txt (hunk 3 above)
Comment (rustagir) [inline docs/query-builder.txt]: S: describe what the plain date meaning of this is

**Review by rustagir — APPROVED** [review]
your changes really cleared things up! left some small suggestions

File: docs/includes/query-builder/QueryBuilderTest.php (hunk 2 above)
Comment (norareidy) [reply docs/includes/query-builder/QueryBuilderTest.php]: I added back the old example and kept this one - lmk what you think

**Review by GromNaN — APPROVED** [review]
LGTM

File: docs/includes/query-builder/QueryBuilderTest.php (hunk 2 above)
Comment (GromNaN) [reply docs/includes/query-builder/QueryBuilderTest.php]: Thanks, that's perfect.

**Pull request merged**